purchase_request/migrations/12.0.1.0.0/post-migration.py

```diff
@@ -0,0 +1,136 @@
+# Copyright 2018-2019 Eficent Business and IT Consulting Services S.L.
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl-3.0).
+
+from openupgradelib import openupgrade
+import logging
+logger = logging.getLogger(__name__)
+
+
+def create_allocation(env, po_line, pr_line, stock_move_id, qty):
+    vals = {'requested_product_uom_qty': qty,
+            'purchase_request_line_id': pr_line,
+            'purchase_line_id': po_line,
+            'stock_move_id': stock_move_id,
+            'allocated_product_qty': 0.0,  # this forces recalculate
+            }
+    alloc = env['purchase.request.allocation'].create(vals)
+    return alloc
+
+
+def create_service_allocation(env, po_line, pr_line, qty):
+    vals = {'requested_product_uom_qty': qty,
+            'purchase_request_line_id': pr_line,
+            'purchase_line_id': po_line,
+            }
+    alloc = env['purchase.request.allocation'].create(vals)
+    return alloc
+
+
+def allocate_stockable(ml, ml_done=None):
+    #  done here because open_product_qty is zero so cannot call method in
+    #  stock_move_line
+    if ml_done is None:
+        ml_done = []
+    ml.product_uom_id._compute_quantity(
+        ml.qty_done, ml.product_id.uom_id)
+    to_allocate_qty = ml.qty_done
+    for allocation in \
+            ml.filtered(
+                lambda m: m.id not in ml_done).move_id.\
+            purchase_request_allocation_ids:
+        if to_allocate_qty > 0.0 and \
+                allocation.allocated_product_qty < \
+                allocation.requested_product_uom_qty:
+            allocated_qty = min(
+                allocation.requested_product_uom_qty, to_allocate_qty)
+            allocation.allocated_product_qty += allocated_qty
+            to_allocate_qty -= allocated_qty
+        ml_done.append(ml.id)
+    return ml_done
+
+
+def create_allocations(env):
+    env['purchase.request'].search([('state', '=', 'purchase')])
+    cr = env.cr
+    #  First allocate stockable and consumables
+    logger.info('Allocating purchase request for stockables '
+                'and consumables')
+    cr.execute(
+        """
+        SELECT purchase_request_line_id, purchase_order_line_id, sm.id,
+         sm.product_qty, sm.product_uom_qty, sm.product_qty, prl.product_qty
+        FROM purchase_request_purchase_order_line_rel rel
+        INNER JOIN purchase_request_line prl ON prl.id = 
+        rel.purchase_request_line_id
+        INNER JOIN purchase_order_line pol ON pol.id = 
+        rel.purchase_order_line_id    
+        INNER JOIN product_product pp ON prl.product_id = pp.id
+        INNER JOIN product_template pt ON pp.product_tmpl_id = pt.id
+        LEFT JOIN stock_move sm ON sm.purchase_line_id = pol.id
+        WHERE pt.type != 'service'
```

Comment:
`AND prl.product_qty - prl.qty_done != 0.0`

Reply:
will not work qty_done is always 0 at the beginning